ProcessLib/ComponentTransport/ComponentTransportFEM.h

```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
+                               velocity_magnitude) *
+                              I +
+                          (solute_dispersivity_longitudinal -
+                           solute_dispersivity_transverse) /
+                              velocity_magnitude * velocity *
+                              velocity.transpose())
+                    : GlobalDimMatrixType(
+                          (porosity * molecular_diffusion_coefficient +
+                           solute_dispersivity_transverse *
+                               velocity_magnitude) *
+                          I);
+
+            // matrix assembly
+            local_M.noalias() += w * N.transpose() * retardation_factor *
+                                     porosity * density * N +
+                                 w * N.transpose() * retardation_factor *
+                                     porosity * C_int_pt * drho_dC * N;
+
+            // coupling term
+            {
+                double p0_int_pt = 0.0;
+                NumLib::shapeFunctionInterpolate(local_p0, N, p0_int_pt);
+
+                local_K.noalias() +=
+                    (-dNdx.transpose() * velocity * density * N +
+                     dNdx.transpose() * density * hydrodynamic_dispersion *
+                         dNdx +
+                     N.transpose() * decay_rate * porosity *
+                         retardation_factor * density * N +
+                     N.transpose() * porosity * retardation_factor * drho_dp *
+                         (p_int_pt - p0_int_pt) / dt * N) *
+                    w;
+            }
+        }
+    }
+
     std::vector<double> const& getIntPtDarcyVelocity(
         const double t,
         GlobalVector const& current_solution,
         NumLib::LocalToGlobalIndexMap const& dof_table,
         std::vector<double>& cache) const override
     {
-        auto const n_integration_points =
-            _integration_method.getNumberOfPoints();
-
         auto const indices = NumLib::getIndices(_element.getID(), dof_table);
         assert(!indices.empty());
-        auto const local_x = current_solution.get(indices);
 
-        // Assuming that fluid density always depends on the concentration of
-        // the component placed at the first.
-        auto const concentration_index = 1 * ShapeFunction::NPOINTS;
-        // get local_C and local_p
-        auto const C_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[concentration_index], concentration_size);
-        auto const p_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[pressure_index], pressure_size);
+        if (_coupled_solutions == nullptr)  // monolithic scheme
+        {
+            auto const local_x = current_solution.get(indices);
+
+            auto const local_p = Eigen::Map<const NodalVectorType>(
+                &local_x[pressure_index], pressure_size);
+            auto concentration_index = 1 * concentration_size;
```

Comment:
`1 * concentration_size`?

Reply:
changed to `auto const concentration_index = 1 * ShapeFunction::NPOINTS;`